### ISSUE TITLE
Add a safe wrapper around `ptrEq`

### ISSUE THREAD

**Issue by gebner** [leanprover/lean4#1502]
We want to use pointer equality as a performance optimization.  For example, as a first check in `DecidableEq` instances: if the two arguments are pointer-equal, then we can immediately return `isTrue`.  However `ptrEq` is unsafe, and every time it is used via `implementedBy` we potentially introduce bugs.  Therefore we want a safe wrapper around `ptrEq`.  There are several issues to consider:

 1. `ptrEq` is not compatible with equality.  `ptrEq a a` will always return true, but `ptrEq a b` may return false even if `a = b`.  This is easy to address by either using a CPS-style definition or wrapping the result in a quotient like `safePtrEq (a b : α) : Squash Bool`.
 2. We need some information about the result of `ptrEq` beyond the fact that it is a Boolean, so that we can conclude `a = b` if `ptrEq a b` returns true.
 3. In general, `ptrEq a b` does not imply `a = b`.

Number (3) causes the `withPtrEq` function in core to be unsafe:
```lean
withPtrEq {α : Type u} (a b : α) (k : Unit → Bool) (h : a = b → k () = true) : Bool
```

The problem is that it is very easy to construct provably different Lean terms that evaluate to identical VM objects.  In hindsight, this is an obvious effect of erasure and common subexpression elimination.  The following are some concrete examples:
```lean
@[simp] theorem Bool_ne_Empty : Bool ≠ Empty :=
  (Empty.rec <| · ▸ true)

-- 1) Types are erased, so ptrEq will always claim that they are equal:
def problem1 := withPtrEq Bool Empty (fun _ => false) (by simp)
#reduce problem1 -- false
#eval problem1 -- true

-- 2) This works even if we're not directly comparing types, but also when
-- they're nested in a heap value (due to CSE).
def problem2 := withPtrEq (Bool, 1) (Empty, 1) (fun _ => false) (by simp)
#reduce problem2 -- false
#eval problem2 -- true
```

Potential fixes:
1. Cripple the compiler to never do CSE after erasure, and have `withPtrEq` return false if one of the arguments is `box(0)`.  (Only mentioned for completeness, I don't think this is a good idea.)
2. Add a type class `class InjectiveVMRepr (α : Sort u) : Prop where private mk ::`, auto-generate instances for inductives and structures that are safe to compare with `ptrEq`, and add an additional `[InjectiveVMRepr α]` argument to `withPtrEq`.  Unfortunately it is not possible to use `unsafe` for such Rust-style "marker traits", because we need to prove `Nonempty (InjectiveVMRepr α)` to define the safe instances.
3. Add a type class `SafePtrEq` with a `safePtrEq (a b : α) : Squash { b : Bool // b → a = b }` operation.  This has the downside of increasing the compiled code size (due to monomorphization), and we can't define instances like `[SafePtrEq α] → SafePtrEq (Array α)`.

**Comment by leodemoura** [comment]
Here is another potential fix:
- Do not inline `ptrEq` at the initial compilation phases. Remark: in the new compiler stack, we are adding support for controlling in which phase a definition is inlined. 
- During code specialization, monomorphise `ptrEq`. 
- When we introduce impure features (e.g., `inc`, `dec`, support for `IOBase`), we also process `ptrEq`. In the new compiler stack, we are preserving type information, the information is sufficient for deciding whether the runtime representation is injective. If it is not `ptrEq a b` is replaced with `false`. Otherwise, we inline it.

**Comment by gebner** [comment]
> During code specialization, monomorphise `ptrEq`.

I like this approach.  It also helps with decidability instances for polymorphic types:
```lean
instance [DecidableEq α] : DecidableEq (List α) := ...
```
With the TC approach, we'd need an extra `[SafePtrEq α]` argument here so that the instance can be optimized when `α` has an injective VM representation.

The only downside is that monomorphization is limited to type-class arguments at the moment.  So a function like `def foo (a b : α) := ptrEq a b` would not be monomorphized (even though that would be necessary to take advantage of the optimization).

**Comment by leodemoura** [comment]
> The only downside is that monomorphization is limited to type-class arguments at the moment. So a function like def foo (a b : α) := ptrEq a b would not be monomorphized (even though that would be necessary to take advantage of the optimization).

This is true, but we are currently working on the new code generator, and we can add support for this feature.

**Comment by Rob23oba** [comment]
@hargoniX Just mentioning this again because it's the only soundness problem of `native_decide` I'm aware of, excluding the obvious ones like `implemented_by` (to be precise I mean the fact that `withPtrEq` as we have it is unsound). It's clear that there is no easy way to prevent the exploits possible through `withPtrEq` without requiring some kind of type class restriction and it seems like the simplest and most natural one to use here would be requring a computable `DecidableEq` instance. Here a proof of concept:
```lean-4
@[noinline]
unsafe def ptrAddrUnsafeWrapper {α : Type u} (x : DecidableEq α × α) : USize :=
  ptrAddrUnsafe x.2

@[inline]
unsafe def ptrEqSafeImpl {α : Type u} [DecidableEq α] (a b : α) : Squash { x : Bool // x → a = b } :=
  if ptrAddrUnsafeWrapper (‹_›, a) = ptrAddrUnsafeWrapper (‹_›, b) then
    Squash.mk ⟨true, lcProof⟩
  else
    Squash.mk ⟨false, nofun⟩

@[implemented_by ptrEqSafeImpl]
def ptrEqSafe {α : Type u} [DecidableEq α] (a b : α) : Squash { x : Bool // x → a = b } :=
  Squash.mk ⟨false, nofun⟩

def foo := ptrEqSafe false true

open Classical
def bar := ptrEqSafe False True -- fails
```
Of course, the compilation is not quite ideal here (in particular, we'd want `DecidableEq` to be erased after we had the `noncomputable` check)

**Comment by datokrat** [comment]
@Rob23oba What are you trying to ensure by requiring a computable `DecidableEq` instance? Is it a problem for your proposal that every type has a "computable" `DecidableEq` instance like this?

```lean
partial def nonterminatingDecidable {P : Prop} : Decidable P :=
  nonterminatingDecidable

instance : DecidableEq Prop := fun _ _ => nonterminatingDecidable
```

**Comment by Rob23oba** [comment]
Right, I guess I underestimated how hard it would be to actually enforce that the type is "good" for pointer equality. Hmm, so `withPtrEqDecEq` is unsafe too?!
```lean-4
def Classical.computableChoice (h : Nonempty α) : α :=
  Classical.computableChoice h
partial_fixpoint

instance (a b : Prop) : Decidable (a = b) :=
  withPtrEqDecEq a b (fun _ => Classical.computableChoice inferInstance)

example : False = True := by native_decide
```
I guess the takeaway here is that you can't have a `withPtrEq` function without special compiler support that works on a generic type? So you can only have a `withPtrEq` function with compiler support that works for types that are completely known to be "safe" (i.e. don't store `Sort u`s)?
